Framework/src/AggregatorRunner.cxx

```diff
@@ -0,0 +1,227 @@
+#include <Monitoring/Monitoring.h>
+#include <Framework/InputRecordWalker.h>
+// QC
+#include "QualityControl/DatabaseFactory.h"
+#include "QualityControl/QcInfoLogger.h"
+#include "QualityControl/ServiceDiscovery.h"
+#include "QualityControl/Aggregator.h"
+
+using namespace AliceO2::Common;
+using namespace AliceO2::InfoLogger;
+using namespace o2::framework;
+using namespace o2::configuration;
+using namespace o2::quality_control::core;
+using namespace o2::quality_control::repository;
+using namespace std;
+using namespace o2::monitoring;
+
+const auto current_diagnostic = boost::current_exception_diagnostic_information;
+
+namespace o2::quality_control::checker
+{
+
+AggregatorRunner::AggregatorRunner(const std::string& configurationSource, const vector<framework::OutputSpec> checkRunnerOutputs)
+  : mDeviceName(createAggregatorRunnerName()),
+    mTotalNumberObjectsReceived(0)
+{
+  try {
+    mConfigFile = ConfigurationFactory::getConfiguration(configurationSource);
+  } catch (...) {
+    // catch the exceptions and print it (the ultimate caller might not know how to display it)
+    ILOG(Fatal) << "Unexpected exception during initialization:\n"
+                << boost::current_exception_diagnostic_information(true) << ENDM;
+    throw;
+  }
+
+  // prepare list of all inputs
+  int i = 0;
+  for (const auto& spec : checkRunnerOutputs) {
+    auto input = DataSpecUtils::matchingInput(spec);
+    input.binding = "checkerOutput" + to_string(i);
+    mInputs.emplace_back(input);
+    i++;
+  }
+}
+
+AggregatorRunner::~AggregatorRunner()
+{
+  if (mServiceDiscovery != nullptr) {
+    mServiceDiscovery->deregister();
+  }
+}
+
+header::DataDescription AggregatorRunner::createAggregatorRunnerDataDescription(const std::string& aggregatorName)
+{
+  if (aggregatorName.empty()) {
+    BOOST_THROW_EXCEPTION(FatalException() << errinfo_details("Empty taskName for task's data description"));
+  }
+  o2::header::DataDescription description;
+  description.runtimeInit(std::string(aggregatorName.substr(0, header::DataDescription::size)).c_str());
+  return description;
+}
+
+std::string AggregatorRunner::createAggregatorRunnerName()
+{
+  return AggregatorRunner::createAggregatorRunnerIdString(); // there is only one thus we can just take the idString
+}
+
+void AggregatorRunner::init(framework::InitContext&)
+{
+  try {
+    initDatabase();
+    initMonitoring();
+    initServiceDiscovery();
+    initAggregators();
+  } catch (...) {
+    // catch the exceptions and print it (the ultimate caller might not know how to display it)
+    ILOG(Fatal) << "Unexpected exception during initialization:\n"
+                << current_diagnostic(true) << ENDM;
+    throw;
+  }
+}
+
+void AggregatorRunner::run(framework::ProcessingContext& ctx)
+{
+  framework::InputRecord& inputs = ctx.inputs();
+  for (auto const& ref : InputRecordWalker(inputs)) { // InputRecordWalker because the output of CheckRunner can be multi-part
+    if (ref.header != nullptr) {
```

Comment:
InputRecordWalker should walk only on non-empty inputs, I think it is redundant.